File: backend/src/main/java/com/gaia/service/UserService.java
package com.gaia.service;

import com.gaia.dto.user.UpdateUserProfileRequest;
import com.gaia.dto.user.UpdateUserRoleRequest;
import com.gaia.dto.user.UserResponse;
import java.util.List;

public interface UserService {

    UserResponse me(String email);

    UserResponse updateMe(UpdateUserProfileRequest request, String email);

    List<UserResponse> findAll();

    UserResponse updateRole(Long id, UpdateUserRoleRequest request, String authenticatedEmail);

    void delete(Long id, String authenticatedEmail);
}

File: frontend/src/app/services/user.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { UpdateUserProfileRequest, UpdateUserRoleRequest, User } from '../models/user.model';

@Injectable({ providedIn: 'root' })
export class UserService {
  private readonly apiUrl = `${environment.apiUrl}/users`;

  constructor(private readonly http: HttpClient) {}

  me(): Observable<User> {
    return this.http.get<User>(`${this.apiUrl}/me`);
  }

  updateMe(request: UpdateUserProfileRequest): Observable<User> {
    return this.http.put<User>(`${this.apiUrl}/me`, request);
  }

  getUsers(): Observable<User[]> {
    return this.http.get<User[]>(this.apiUrl);
  }

  updateRole(id: number, request: UpdateUserRoleRequest): Observable<User> {
    return this.http.put<User>(`${this.apiUrl}/${id}/role`, request);
  }

  deleteUser(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}
